File: src/java/sorts/QuickSort.java
package java.sorts;

/**
 * @fileName: QuickSort.java
 * @author: wyli
 * @createTime: 2023/9/3 21:19
 * @description:
 */
public class QuickSort {

    public static void quickSort(int[] a, int n){
        quickInternallySort(a, 0, n - 1);
    }
    public static void quickInternallySort(int[] a, int p, int r){
        if (p >= r){
            return;
        }
        int q = partition(a, p,  r);
        quickInternallySort(a, p, q - 1);
        quickInternallySort(a, q + 1, r);
    }
    public static int partition(int[] a, int p, int r){
        int pivot = a[r];
        int i = p;
        for (int j = p; j < r; j++) {
            if (a[j] < pivot){
                if (i == j){
                    i ++;
                }else{
                    int tmp = a[i];
                    a[i ++] = a[j];
                    a[j] = tmp;
                }
            }
        }
        int tmp = a[i];
        a[i] = a[r];
        a[r] = tmp;
        return i;
    }
}

File: src/cpp/sorts/QuickSort.cpp
#include<iostream>
#include<vector>


class QuickSort{
public:
    void quickSort(std::vector<int>& arr){
        int len = arr.size();
        quickInternallySort(arr, 0, len - 1);
    }
    
    void quickInternallySort(std::vector<int>& arr, int i, int j){
        if(i >= j){
            return;
        }
        int pos = partition(arr, i, j);
        quickInternallySort(arr, i, pos - 1);
        quickInternallySort(arr, pos + 1, j);
    }

    int partition(std::vector<int>& arr, int i, int j){
        int val = arr[j];
        int l = i;
        int k = i;
        for (; k < j; k++)
        {
            if (arr[k] < val)
            {
                if (k == l)
                {
                    l++;
                }else{
                    int tmp = arr[k];
                    arr[k] = arr[l];
                    arr[l ++] = tmp;
                }
            }
        }
        arr[j] = arr[l];
        arr[l] = val;
        return l;
    }
};
